### PR TITLE
Katharopoulos2020transformers kelseyball

### PR DESCRIPTION
**Pull request by kelseyball**
Hey Sai, my summary is ready for review. Sorry for the delay.

### CODE REVIEW

File: _summaries/katharopoulos2020transformers_1.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+layout: summary
+title: Summary
+paper: katharopoulos2020transformers
+# Please fill out info below
+author: kelseyball
+score: 9
+---
+
+# Core Idea
+Expressing self-attention using kernel feature maps can reduce its time and space complexity from quadratic to linear
```
Comment (saikm200022) [inline _summaries/katharopoulos2020transformers_1.md:11]: Another core component to their paper was re-expressing their transformer model as a RNN to achieve much greater inference speeds.

File: _summaries/katharopoulos2020transformers_1.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+The associative property of matrix multiplication allows us to make the following simplification,
+
+<img width="500px" src="katharopoulos2020transformers_4a.png"/>
+
+The reduction from quadratic to linear complexity comes from the fact that the matrix product of the keys and values can be computed once and reused for every query.
+
+# Linear Autoregressive Transformers
+- In autoregressive tasks, inputs from future timesteps are masked to the model. For autoregressive transformers, we cannot precompute the full value of the key-value matrix product as shown above.
+- However, the authors show that part of the attention computation at each timestep can be computed iteratively. In other words, it can be computed in constant time using the previous value. This observation allows for time complexity linear in the sequence length.
```
Comment (saikm200022) [inline _summaries/katharopoulos2020transformers_1.md:40]: I think you should bring up these two points for some context. With autoregressive transformers, it is easier to parallelize during training and not possible during test time. The attention computation must also be repeated for previous time steps.

Comment (saikm200022) [inline _summaries/katharopoulos2020transformers_1.md:40]: I would be a little bit more specific when you are talking about what part of the attention computation that is being stored and updated as internal state: phi (K_j ) V^T

File: _summaries/katharopoulos2020transformers_1.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+- Similarly, gradients can be derived iteratively, enabling both the forward and backward pass to be computed in linear time and constant memory.
+
+# Autoregressive Transformers as RNNs
+- The recurrence noted in the previous section allows the authors to express an autoregressive transformer as an RNN:
+<img width="500px" src="katharopoulos2020transformers_5a.png"/>
+- "Best of bost worlds": linear, autoregressive transformers can be parallelized (unlike RNNs) and also have linear-in-the-sequence time complexity (unlike traditional softmax attention)
+
+# Experiments
+- **Synthetic experiments:** They train autoregressive transformers for an artificial copy task. They show that computation and memory scales linearly with linear transformer and reformer, and quadratically with softmax attention.
```
Comment (saikm200022) [inline _summaries/katharopoulos2020transformers_1.md:49]: The paper has 2 graphs that are side-by-side of time and GPU memory usage vs. Sequence Length for the different architectures tested. Adding that image will bolster your summary of the synthetic experiments.

File: _summaries/katharopoulos2020transformers_1.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+- In autoregressive tasks, inputs from future timesteps are masked to the model. For autoregressive transformers, we cannot precompute the full value of the key-value matrix product as shown above.
+- However, the authors show that part of the attention computation at each timestep can be computed iteratively. In other words, it can be computed in constant time using the previous value. This observation allows for time complexity linear in the sequence length.
+- Similarly, gradients can be derived iteratively, enabling both the forward and backward pass to be computed in linear time and constant memory.
+
+# Autoregressive Transformers as RNNs
+- The recurrence noted in the previous section allows the authors to express an autoregressive transformer as an RNN:
+<img width="500px" src="katharopoulos2020transformers_5a.png"/>
+- "Best of bost worlds": linear, autoregressive transformers can be parallelized (unlike RNNs) and also have linear-in-the-sequence time complexity (unlike traditional softmax attention)
+
+# Experiments
```
Comment (saikm200022) [inline _summaries/katharopoulos2020transformers_1.md:48]: I would mention somewhere in the Experiments section that despite performing much faster, "Linear Transformers" attained the same final performance as the traditional softmax transformer.

File: _summaries/katharopoulos2020transformers_1.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+- **Synthetic experiments:** They train autoregressive transformers for an artificial copy task. They show that computation and memory scales linearly with linear transformer and reformer, and quadratically with softmax attention.
+- **Image generation:** They use autoregressive transformers to generate CIFAR-10 images pixel by pixel. Linear transformer generates images 4000x times faster than transformer with softmax attention.
```
Comment (saikm200022) [inline _summaries/katharopoulos2020transformers_1.md:50]: Including table 2 from the paper can also provide some perspective on how the other methods performed on this particular task.

**Review by saikm200022 — CHANGES_REQUESTED** [review]
Great, concise summary! I have some suggestions that I believe can improve your summary.

**Review by zhaoyue-zephyrus — APPROVED** [review]
